### ISSUE TITLE
Overly long TXT record breaks lmdbbackend import

### ISSUE THREAD

**Issue by mzealey** [PowerDNS/pdns#8012]
- Program: Authoritative
 - Issue type: Bug report

### Short description

Long TXT records cause lmdb import process to break with the error message:

```
$ pdnsutil load-zone example.com example.com.db
Error: putting data: MDB_BAD_VALSIZE: Unsupported size of key/DB name/data, or wrong DUPFIXED size
```

Example zone file `example.com.db`:

```
example.com.    86400   IN      SOA     example.com.    example.com.    (
                                                2018031306 ;Serial Number
                                                14400 ;refresh
                                                7200 ;retry
                                                3600000 ;expire
                                                14400   )

example.com.    10      IN      TXT     "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
```

I guess you are storing the TXT aspect of the detail in the lmdb key record which causes it to go over the 512 byte key limit? If this is the case it may need a db design change before launching the new lmdb backend as I suspect TXT records should be stored as values.

### Environment
 - Operating system: centos7
 - Software version: pdns-4.2.0-rc2.169.master.g85474361f-centos-7.tar.bz2

**Comment by Habbie** [comment]
We appear to be using DUPSORT which makes the value participate in the 512 byte limit, indeed.

**Comment by mzealey** [comment]
I'm not 100% sure of the physicalities of the underlying DNS packets but I think that a TXT record can not be over 255 bytes (are other records allowed to be longer?). This *may* mean that this is not an issue and you should expect input data that is accordingly valid somehow... Either way I think this needs a better error message ie marking the zone as invalid in the bind backend (or elsewhere in the system).

**Comment by Habbie** [comment]
>  I think that a TXT record can not be over 255 bytes (are other records allowed to be longer?).

TXT can be up to ~64K (but anything beyond a kilobyte or two is unrealistic for practical uses). Many other records can also be >255 but most have no business being over a few hundred bytes.

**Comment by mzealey** [comment]
You may need to look at having a break-out table for storing longer records in that case then :-/

**Comment by gibson042** [comment]
TXT RDATA can be up to 64 KB, but it is [composed of \<character-string>s](https://tools.ietf.org/html/rfc1035#section-3.3.14), each of which is itself [limited to 256 bytes](https://tools.ietf.org/html/rfc1035#section-3.3) (a single length octet followed by up to 255 octets of contents). The sample zone file includes an unacceptable 600-character \<character-string>.

**Invalid (too many characters in \<character-string> contents)**
```
example.com.    10      IN      TXT     "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
```

**Valid (multiple \<character-string>s, each with contents not exceeding 255 characters)**
```
example.com.    10      IN      TXT    (
  "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
  "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
  "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
  "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
  "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
  "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
)
```

**Comment by Habbie** [comment]
> **Invalid (too many characters in <character-string> contents)**

While invalid in formal syntax, it has been legal in PowerDNS (in all backends, including the bindbackend) for decades, and PowerDNS will automatically split it into chunks.

**Comment by Habbie** [comment]
It looks like we might not fix this for 4.2; what we need to do for 4.2 is add a schema version to the 'main' LMDB so we can do migrations later.

**Comment by Habbie** [comment]
4.2 now has schema versioning (once #8165 is merged). Moving this to milestone 4.3.0.

**Comment by peterthomassen** [comment]
Just ran into this as well. I would be very grateful if this could be addressed for 4.3.0!

**Comment by Habbie** [comment]
> Just ran into this as well. I would be very grateful if this could be addressed for 4.3.0!

This is unlikely, sorry :(

**Comment by My1** [comment]
can this be worked around, e.g. by splitting the records manually or whatever?

**Comment by Habbie** [comment]
> can this be worked around, e.g. by splitting the records manually or whatever?

Only if whatever needs the TXT records accepts that; or, more generally put: no, sorry.

**Comment by My1** [comment]
okay sad. I hope this can be fixed soon-ish too, as I kinda need to ax my dkim/dmarc stuff, AWESOME (not)

**Comment by NewRedsquare** [comment]
Yup, i can't add a 4096bits OPENPGPKEY record

**Comment by RobinGeuze** [comment]
So I poked at this a bit.

Adding a new schema-version that stores all rrset's rather than rr's is relatively easy. Conversion logic is doable as well, and you could even backup the old shards.

However, you do run into a limitation of dnsbackend/ueberbackend as it is now. "get" returns a single rr at a time, which means that if you load an entire rrset at once you need to temporarily store the rest of the rrset you cannot immediately return. The other approach would be a major refactor where we have get return an RRSet instead of a DNSResourceRecord. Looking at the code that could work, but it would be a major refactor since multiple backends would need to be refactored (for example the remotebackend).

**Comment by Habbie** [comment]
> get" returns a single rr at a time, which means that if you load an entire rrset at once you need to temporarily store the rest of the rrset you cannot immediately return.

Yes, this makes sense, and should not be very hard. I vaguely recall that one or two backends already do it.

> The other approach would be a major refactor where we have get return an RRSet instead of a DNSResourceRecord.

That would make a lot more sense than the current interface, yes. It's been on my list for a while.

We could have `getRRSet` in DNSBackend, inherited by the backends that only supply `get`. This came up recently in another discussion as well, but I can't immediately find it.

For both cases, note that `get` will never return a partial RRset, but it may return -multiple- RRsets (in fact, it usually does, because we ask the backend for ANY).

**Comment by Habbie** [comment]
> For both cases, note that `get` will never return a partial RRset, but it may return -multiple- RRsets (in fact, it usually does, because we ask the backend for ANY).

Which I guess means that the obvious refactor is 'vector of RRsets', not 'one RRset'.

**Comment by rgacogne** [comment]
> We could have getRRSet in DNSBackend, inherited by the backends that only supply get. This came up recently in another discussion as well, but I can't immediately find it.

I wrote a PoC doing something like that in https://github.com/PowerDNS/pdns/compare/master...rgacogne:auth-getbestauth3 (the code is ugly at the moment, please brace yourself!), perhaps that's what you had in mind?
